### ISSUE TITLE
Contents of verification-metadata.xml depend on whether dependencies are cached

### ISSUE THREAD

**Issue by akwizgran** [gradle/gradle#19228]
### Expected Behavior

Running `./gradlew --write-verification-metadata sha256 <task>` (without an existing `verification-metadata.xml` file) should record the same dependency checksums in `verification-metadata.xml` regardless of whether the dependencies are found in the cache or not.

### Current Behavior

For certain dependencies, running `./gradlew --write-verification-metadata sha256 <task>` (without an existing `verification-metadata.xml` file) records a larger set of dependency checksums when running with an empty cache than when running with a populated cache.

Specifically, the .pom files for certain dependencies (e.g. `org.junit:junit-bom:5.7.0`) are only recorded in `verification-metadata.xml` if the dependencies are **not** found in the cache.

If the `verification-metadata.xml` file is created by a developer who has the dependencies in their local cache, and is then checked into version control, the file will not contain these specific dependencies, which may cause build failures for other developers who don't have those dependencies in their local caches.

### Context

We're trying out Gradle's built-in dependency verification in a new project. When a developer created the `verification-metadata.xml` file locally and checked it into version control, the CI pipeline failed with the following error:

```
> Dependency verification failed for configuration ':dont-kill-me-lib:lintClassPath':
    - On artifact junit-bom-5.7.0.pom (org.junit:junit-bom:5.7.0) in repository 'BintrayJCenter': checksum is missing from verification metadata.
  
  If the artifacts are trustworthy, you will need to update the gradle/verification-metadata.xml file by following the instructions at https://docs.gradle.org/7.3/userguide/dependency_verification.html#sec:troubleshooting-verification
  
  These files failed verification:
    - GRADLE_USER_HOME/caches/modules-2/files-2.1/org.junit/junit-bom/5.7.0/6c667331d41729ca3c8e02d1c26ae1334a81357f/junit-bom-5.7.0.pom
  
  GRADLE_USER_HOME = /builds/briar/briar-mailbox/.gradle
  
  These files failed verification:
    - GRADLE_USER_HOME/caches/modules-2/files-2.1/org.junit/junit-bom/5.7.0/6c667331d41729ca3c8e02d1c26ae1334a81357f/junit-bom-5.7.0.pom
```

We investigated this error and found that this dependency was not recorded by `--write-verification-metadata` on the developer's machine when using the default `GRADLE_USER_HOME`, but setting `GRADLE_USER_HOME` to an empty directory, which caused all dependencies to be downloaded, caused `--write-verification-metadata` to record this dependency.

### Steps to Reproduce 

The following repo contains a minimal example project to reproduce the issue:

https://code.briarproject.org/akwizgran/dependency-verification

The `reproduce-bug.sh` script creates a temporary `GRADLE_USER_HOME` and writes the `verification-metadata.xml` file twice. The first time the cache is empty, the second time it contains the artifacts downloaded the first time. The script prints the diff between the metadata files (which should be empty, but isn't).

The script then stops the Gradle daemon, deletes the temporary `GRADLE_USER_HOME` to clear the cache, and runs the same Gradle task that was used to generate the metadata files. This task fails because the cache is now empty, so the build finds the `junit-bom-5.7.0.pom` dependency that's missing from `verification-metadata.xml`.

The example project uses the Android library plugin. The repo also contains a [java-plugin branch](https://code.briarproject.org/akwizgran/dependency-verification/-/tree/java-plugin) that uses the built-in Java plugin instead. The `reproduce-bug.sh` script is the same. When using the Java plugin, the contents of `verification-metadata.xml` still differ depending on whether the dependencies are cached, but strangely the final task does **not** fail due to extra dependencies being found, as it does when using the Android library plugin.

I believe the bug is in Gradle rather than the Android plugin because the contents of `verification-metadata.xml` differ depending on whether the dependencies are cached, regardless of whether the Android library plugin or the Java plugin is used. But I don't understand why this causes the final task to fail with the Android library plugin but not with the Java plugin.

### Your Environment

* Gradle version: 7.3
* Android plugin version: 7.0.3
* OS: Debian buster

**Comment by jbartok** [comment]
Might be related to https://github.com/gradle/gradle/issues/19126

**Comment by ljacomet** [comment]
This is most likely caused by the way Gradle will search for the module file when the POM contains a specific marker. Once cached, that information is most likely not encoded and thus the POM is never touched anymore. However on a new machine, the POM is downloaded once indeed.

Can you confirm that the generated incomplete file contains entries for `.module` ?

**Comment by akwizgran** [comment]
@ljacomet yes, the incomplete and complete files both contain `junit-bom-5.7.0.module`, but only the complete file contains the corresponding POM.

**Comment by yogurtearl** [comment]
This is an issue, please keep it open.

**Comment by liutikas** [comment]
Any updates on this issue?

**Comment by eighthave** [comment]
Looks like this issue strikes again with Gradle 8.10.2
https://github.com/guardianproject/ripple/actions/runs/11262630001/job/31318900634

**Comment by nikclayton** [comment]
Still an issue with 8.14

```
------------------------------------------------------------
Gradle 8.14
------------------------------------------------------------

Build time:    2025-04-25 09:29:08 UTC
Revision:      34c560e3be961658a6fbcd7170ec2443a228b109

Kotlin:        2.0.21
Groovy:        3.0.24
Ant:           Apache Ant(TM) version 1.10.15 compiled on August 25 2024
Launcher JVM:  21.0.6 (JetBrains s.r.o. 21.0.6+-13355223-b631.42)
Daemon JVM:    /home/nik/android-studio/jbr (no JDK specified, using current Java home)
OS:            Linux 6.11.0-24-generic amd64
```

@ov7a I appreciate the team may not have time to expedite a fix, but could the documentation at https://docs.gradle.org/8.14/userguide/dependency_verification.html be updated to reference this problem, with all examples modified to use an empty GRADLE_USER_HOME?

**Comment by big-guy** [comment]
@ljacomet do you know if `--refresh-dependencies` helps here or does it only work properly with a clean GRADLE_USER_HOME?

We won't be able to address this for 9.0, but we can update docs for troubleshooting for now (#33500) and our intention would be that you can always reliably generate dependency verification metadata regardless of cache state.

**Comment by yogurtearl** [comment]
`--write-verification-metadata` is still non-deterministic for any "BOM" poms or transitive deps of bom poms, sometimes it adds the needed info, and sometime it omits the info. 

If I pre-populate all the BOM info into `verification-metadata.xml` before I do `--write-verification-metadata` to update it, then it won't remove the pre-populated info. 

Of course, this is not great as it requires us to manually curate a list of all our bom poms and their transitive deps.

**Comment by derekslenk** [comment]
A data point that may help move this: we hit this **systematically** rather than intermittently, and isolated it with a controlled comparison.

Setup: self-hosted Renovate processing 15 Gradle repositories through a single `GRADLE_USER_HOME` per run. That means every repository after the first regenerates its `verification-metadata.xml` against a warm dependency metadata cache — so the incompleteness this issue describes isn't an occasional annoyance for us; it happens on essentially every automated regeneration, and the written file reliably omits parent-POM / platform-BOM entries. The fingerprint is always the same: the `.module` entry is recorded, the parent/BOM `.pom` is missing, and consumer builds fail dependency verification on exactly those.

Controlled comparison against a real incomplete file with five known-missing entries (each arm reset to the same committed baseline and asserted to start at 0 of 5; warm arms restored from one snapshotted `GRADLE_USER_HOME`; Gradle 9.6.1):

| Arm | Cache | Added (of 5 missing) |
|---|---|---|
| `--write-verification-metadata sha256 dependencies` | warm | **0** |
| same + `--refresh-dependencies` | warm | **5** |
| `--write-verification-metadata sha256 dependencies` | cold | 5 |
| `--write-verification-metadata sha256 build` | warm | **0** |

Two conclusions from the arms: cache state is the whole variable (cold run is complete with no other change), and the task following the flag is irrelevant — `build` on an equally warm cache omits the same entries, which is worth noting because task-swapping is the folk fix that keeps circulating (we fell for it too, twice, before measuring).

Workaround that holds up in production: append `--refresh-dependencies` to the write command. We've been running it across the 15-repo fleet; every regeneration since has produced complete files, at the cost of re-fetching descriptors on each run.
